lib/galaxy/webapps/galaxy/services/library_contents.py

```diff
@@ -0,0 +1,296 @@
+            )
+            assoc = trans.app.model.ImplicitlyConvertedDatasetAssociation(
+                parent=content.library_dataset_dataset_association,
+                dataset=content_conv.library_dataset_dataset_association,
+                file_type=content_conv.library_dataset_dataset_association.extension,
+                metadata_safe=True,
+            )
+            trans.sa_session.add(assoc)
+            with transaction(trans.sa_session):
+                trans.sa_session.commit()
+
+    def delete(
+        self,
+        trans: ProvidesHistoryContext,
+        id: DecodedDatabaseIdField,
+        payload: LibraryContentsDeletePayload,
+    ) -> LibraryContentsDeleteResponse:
+        """Delete the LibraryDataset with the given ``id``."""
+        rval = {"id": id}
+        try:
+            ld = self.get_library_dataset(trans, id, check_ownership=False, check_accessible=True)
+            user_is_admin = trans.user_is_admin
+            can_modify = trans.app.security_agent.can_modify_library_item(trans.user.all_roles(), ld)
+            log.debug("is_admin: %s, can_modify: %s", user_is_admin, can_modify)
+            if not (user_is_admin or can_modify):
+                raise exceptions.InsufficientPermissionsException(
+                    "Unauthorized to delete or purge this library dataset"
+                )
+
+            ld.deleted = True
+            if payload.purge:
+                ld.purged = True
+                trans.sa_session.add(ld)
+                with transaction(trans.sa_session):
+                    trans.sa_session.commit()
+
+                # TODO: had to change this up a bit from Dataset.user_can_purge
+                dataset = ld.library_dataset_dataset_association.dataset
+                no_history_assoc = len(dataset.history_associations) == len(dataset.purged_history_associations)
+                no_library_assoc = dataset.library_associations == [ld.library_dataset_dataset_association]
+                can_purge_dataset = not dataset.purged and no_history_assoc and no_library_assoc
+
+                if can_purge_dataset:
+                    try:
+                        ld.library_dataset_dataset_association.dataset.full_delete()
+                        trans.sa_session.add(ld.dataset)
+                    except Exception:
+                        pass
+                    # flush now to preserve deleted state in case of later interruption
+                    with transaction(trans.sa_session):
+                        trans.sa_session.commit()
+                rval["purged"] = True
+            with transaction(trans.sa_session):
+                trans.sa_session.commit()
+            rval["deleted"] = True
+        except Exception as exc:
+            log.exception(f"library_contents API, delete: uncaught exception: {id}, {payload}")
+            raise exceptions.InternalServerError(util.unicodify(exc))
+        return LibraryContentsDeleteResponse(**rval)
```

Comment:
I would try to move away from this pattern. Either the request succeeded with a status code < 300 or it didn't.